modules/memfault-firmware-sdk/memfault_location_metrics.c

```diff
@@ -0,0 +1,197 @@
+ * @param accuracy_m pointer to location accuracy in meters
+ */
+static void prv_location_method_results_record(enum location_method method,
+					       enum location_event_id id,
+					       const struct location_data_details *details,
+					       const float *accuracy_m)
+{
+	if (method == LOCATION_METHOD_GNSS) {
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_method_time_ms,
+					     details->elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_method_result, id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_on_time_ms,
+					     details->gnss.elapsed_time_gnss);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_satellites_tracked_count,
+					     details->gnss.satellites_tracked);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_satellites_used_count,
+					     details->gnss.satellites_used);
+
+		// Only record TTF and accuracy when a fix is acquired
+		if (id == LOCATION_EVT_LOCATION) {
+			MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_gnss_time_to_fix_ms,
+						     details->gnss.pvt_data.execution_time);
+
+			if (accuracy_m != NULL) {
+				uint32_t accuracy_cm = (uint32_t)(*accuracy_m * 100.0f);
+				MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_accuracy_cm, accuracy_cm);
+			}
+		}
+	}
+#if defined(CONFIG_LOCATION_METHOD_CELLULAR)
+	else if (method == LOCATION_METHOD_CELLULAR) {
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_method_time_ms,
+					     details->elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_method_result, id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_neighbor_cells_count,
+					     details->cellular.ncells_count);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_lte_gci_cells_count,
+					     details->cellular.gci_cells_count);
+	}
+#endif
+#if defined(CONFIG_LOCATION_METHOD_WIFI)
+	else if (method == LOCATION_METHOD_WIFI) {
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_wifi_method_time_ms,
+					     details->elapsed_time_method);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_wifi_method_result, id);
+		MEMFAULT_METRIC_SET_UNSIGNED(ncs_loc_wifi_access_points_count,
+					     details->wifi.ap_count);
+	}
+#endif
+}
+
+/**
+ * @brief Record the fallback method used
+ *
+ * @param method fallback method
+ * @param fallback pointer to fallback method data
+ */
+static void prv_location_method_fallback_record(enum location_method method,
+						const struct location_data_fallback *fallback)
+{
+	prv_location_method_results_record(method, fallback->cause, &fallback->details, NULL);
+
+	// Log the results
+	char method_str[16];
+	snprintk(method_str, sizeof(method_str), "%s", prv_location_method_to_string(method));
+	LOG_DBG("%s method attempted", method_str);
+	char fallback_method_str[16];
+	snprintk(fallback_method_str, sizeof(fallback_method_str), "%s",
+		 prv_location_method_to_string(fallback->next_method));
+	LOG_DBG("Falling back to %s method", fallback_method_str);
+}
+
+/**
+ * @brief Start a location session.
+ *
+ */
+static void prv_location_metrics_session_start(void)
+{
+	if (s_session_in_progress) {
```

Comment:
tiny bit surprised this is required, seems like the `LOCATION_EVT_STARTED` event shouldn't fire more than once- did it hit in testing? no biggie though.

Reply:
Did not hit in testing -- more of a guard to prevent out of order calls to start and stop. Originally I considered making the start/stop public in which case it would be pretty essential to have that guard, but I still think it is a good protection within the module.

Reply:
This shouldn't be needed (but might be good to have as you say), but if it was I'd consider using an atomic variable or a mutex/semaphore. I guess this is always run in `location_core_work_q` which calls the event handler so not a major concern.